Replace isort.sh with isort.py for better logging and CLI argument parsing

Eric-Arellano opened the pull request:
### Problem
`isort.sh` suffers from three problems:

1. When not using `-f`, the output upon failure is unnecessarily wordy and does not say which files were the actual issue. For example:
```
FAILURE: Exited with return code 1 while running `/Users/eric/.pyenv/shims/python3.6 /Users/eric/.cache/pants/python/CPython-3.6.8/d6bcba436bdefbc27a69dd880d6ee5716e004764/isort-c201e856817dc9ae802d18e3ec45a3bfa4c2ac11.pex --check-only src/python/pants/util/tarutil.py src/python/pants/util/retry.py src/python/pants/util/osutil.py src/python/pants/util/memo.py src/python/pants/util/process_handler.py src/python/pants/util/strutil.py src/python/pants/util/py2_compat.py src/python/pants/util/meta.py src/python/pants/util/xml_parser.py src/python/pants/util/objects.py src/python/pants/util/rwbuf.py src/python/pants/util/collections_abc_backport.py`.

FAILURE
```

2. When using `-f`, we swallow all output. Presumably, the user would like to know which files were changed, rather than having to run `git status` to do this.
3. `-f` is a bit cryptic. Would be nice to allow either `-f` or `--fix`.

### Solution
Port the script to modern Python 3, now that we have Python 3 in all CI shards thanks to https://github.com/pantsbuild/pants/pull/7483.

Note the original approach was to improve the actual `fmt.isort` task in https://github.com/pantsbuild/pants/pull/7462, but it was resolved the task could not be converted into a ConsoleTask.

### Result
If `--fix` is passed and files were fixed, we will output:
```
The following files were fixed.

src/python/pants/util/strutil.py
```

If `--fix` is left off, and files had bad sort order, we will output in red:
```
The following files have incorrect import orders. Fix by running `./build-support/isort.py --fix`.

src/python/pants/util/strutil.py
```

If there is an unexpected failure, we will print stderr and fail the script.

cosmicexplorer commented on build-support/bin/isort.py:
```diff
@@ -0,0 +1,70 @@
+    else:
+      # NB: we intentionally don't swallow stderr, so that will be printed before
+      # this message.
+      die("Unexepcted failure.")
```
```suggestion
      die("Unexpected failure.")
```

Eric-Arellano replied:
Good catch, thanks.

cosmicexplorer commented on build-support/bin/common.py:
```diff
@@ -0,0 +1,67 @@
+
+
+def green(message: str) -> None:
+  print(f"{_COLOR_GREEN}{message}{_COLOR_RESET}")
```
I'm fine with leaving these helper methods in this file for now in order to unblock others from jumping in and writing python 3 scripts themselves without having to reinvent the wheel, especially since some of this is used in the setup repo.

Eric-Arellano replied:
Where else would these go? It's meant to emulate `common.sh`.

One limitation, though, is that this file has to live in `build-support/bin` because Python scripts cannot import files from their parent directory, unless we are willing to accept `sys` path hackery at the top of our `bin/*.py` scripts.

jsirois replied:
I'm fine with it living here, but can you kill all the unused code? you only need `die`, `_COLOR_RED` and `_COLOR_RESET` fwict. This is generally not awesome though as a script. We should probably be dogfooding and just using a custom task and re-using ansicolors, etc. Since isort et all currently don't produce data products though that you could consume and format to output, thats not an easy task at the moment.

jsirois replied:
... and towards this end - all this work at making isort useable is lost on Pants users. It wouldn;t be lost though is fixed via dogfooding.

cosmicexplorer commented on build-support/bin/pre-commit.sh:
```diff
@@ -56,8 +56,7 @@ echo "* Checking for bad shell patterns" && ./build-support/bin/check_shell.sh |
 if git rev-parse --verify "master" &>/dev/null; then
-  echo "* Checking imports" && ./build-support/bin/isort.sh || \
-    die "To fix import sort order, run \`\"$(pwd)/build-support/bin/isort.sh\" -f\`"
+  echo "* Checking imports" && ./build-support/bin/isort.py || exit 1
```
So this is a fantastic change...how do you feel about adding a test in https://github.com/pantsbuild/pants/blob/master/tests/python/pants_test/repo_scripts/test_git_hooks.py? It's much easier to do this sort of thing as we refactor the repo scripts, I think. Since we're relying on pants to do changed detection instead of piping in a git diff, we can probably avoid some of the more annoying git repo setup this time.

Eric-Arellano replied:
Will look into how to do this today!

jsirois commented on build-support/bin/isort.py:
```diff
@@ -0,0 +1,70 @@
+
+
+def create_parser() -> argparse.ArgumentParser:
+  parser = argparse.ArgumentParser(description="Run isort over build-support to ensure valid import order.")
```
Erm - more disease here of not just using a task plugin in the source trees we already lint. Having to setup argparse, etc.

stuhood commented:
FTR: this applies to both `isort` and `lint` a few lines down: see https://github.com/pantsbuild/pants/blob/272b0a514619edfc6a7768e52d4d96ca15f1c7f3/build-support/bin/pre-commit.sh#L62-L66 ... it would be great to see whether we can fix the underlying problem.

Eric-Arellano commented:
I'm confused by the feedback on this PR. My original approach was to fix the upstream task with https://github.com/pantsbuild/pants/pull/7462 by making it a ConsoleTask, but then John explained we cannot do that https://github.com/pantsbuild/pants/pull/7462#issuecomment-478202963. So, I improved upstream as much as I knew how to, and made this as the next best fix.

My confusion is where to go from here in addressing the feedback. Do I scrap this and rewrite isort to be a V2 goal? I'm not sure how else we could upstream these change or do things like use ansicolors.

Eric-Arellano commented:
Rewriting like this to V1 or to V2 is outside the scope of what I'm looking to contribute. Happy to help if someone else wants to take this on.
